### DISCUSSION TITLE
Questions, dynamic autocomplete.

### DISCUSSION THREAD

**Discussion by Symbiomatrix** [DominikDoom/a1111-sd-webui-tagcomplete#134] [Q&A] [answered]
Hello. First of all, really fantastic utility that is a great quality of life improvement over the existing interface, I think.
I glanced over the code a bit and pondered a few things (still there):
1) Why does write_tag_base_path write the tags' relative path instead of the absolute path? It breaks when extensions are stored outside the webui directory (as I have, for portability), yet seems to make no difference for the js function that reads it.
2) It goes through some trouble to write the hypernetworks, loras and possibly embeddings (didn't work for me) into text files in a temp directory, but then it doesn't do anything with them in terms of autocomplete, they don't seem to appear - what's the purpose of that? Is it just broken on my end?
3) I think that one could easily & quickly convert the list of loras and embeddings to csv format with an arbitrary high count for ordering (or use some file metadata such as date updated) and either a regular or special type, so that they are much easier to search for whilst typing a prompt - have you given that any thought? I believe such a mechanism alone would be worthy of integration into the main repo.

Edit: Oh, I see - lora and embedding lookup works when typing "<". Clever.

**Comment by DominikDoom** [comment, marked as answer]
#### 1:
Gradio forbids loading files that aren't either in the root folder of the app or put into temp directories by gradio itself for security reasons. That means even if I would use absolute paths there, the javascript will not be able to fetch these files if they aren't under the webui folder. It's a bit of an inconvenience, but nothing I can do about it as far as I'm aware. For the same reason, this also breaks symlinks and some docker setups.

#### 2 & 3:
As you found out yourself already, `<` will show these embeddings, loras and hypernets. It also supports showing only one type by using `<e:`, `<l:` or `<h:` respectively, or the long form `<lora:` and `<hypernet:`. Similarly, wildcards will be shown when typing `__`, if you have any. Embeddings also have the additional functionality of filtering between v1 and v2 Embeddings that work with Stable Diffusion 1.x and 2.x models respectively, by just typing v1/v2 after the < or <e: before typing the actual search term.

The reason they are written to those files in the first place is similar to point one: Gradio doesn't provide a file index or allow searching for local files, so the script needs to find and save the paths beforehand for all files it needs to load in the browser. This applies to all paths that can have variable file or folder names, which is also the reason for tag_base_path to exist, since the extensions folder can be called whatever the user wants.

**Reply by Symbiomatrix** [reply]
Thanks.
1: In that case, perhaps any working files (ie tags subdirectory) could be copied / hardlinked to a fixed temp location under FILE_DIR ahead instead, at least when the extensions directory isn't nested? Bit of a messy workaround leaving loose temp file copies, and I guess permissions might be dodgy, but at least they're not too heavy (certainly no worse than the temp image copies gradio used to produce till that was circumvented a few months back).
2/3: Good to know. Would highly recommend mentioning it in the readme.

**Reply by DominikDoom** [reply]
I might give manual copying a try, but it's definitely a bit dodgy. What would be optimal really would be a way to whitelist specific external paths on Gradio's side to allow loading files from there, but I can understand that they don't have that from a security standpoint.

There was a bit of basic info about embeddings in the readme already, but I've taken you up on your suggestion and properly described the newer additions [here](https://github.com/DominikDoom/a1111-sd-webui-tagcomplete#additional-completion-support). 👍
There's a few other minor features I haven't had the time yet to add to the readme, so I highly recommend browsing the settings to see what's available if you haven't already, or even have a look at [past releases](https://github.com/DominikDoom/a1111-sd-webui-tagcomplete/releases) for more in-depth explanations of some new features.
